airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -0,0 +1,153 @@
+import functools
+from abc import ABC, abstractmethod
+from typing import Any, List, Mapping, Optional, Protocol, Tuple
+
+from airbyte_cdk.sources.connector_state_manager import ConnectorStateManager
+from airbyte_cdk.sources.message import MessageRepository
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+
+
+def _extract_value(mapping: Mapping[str, Any], path: List[str]) -> Any:
+    return functools.reduce(lambda a, b: a[b], path, mapping)
+
+
+class Comparable(Protocol):
+    """Protocol for annotating comparable types."""
+
+    @abstractmethod
+    def __lt__(self: "Comparable", other: "Comparable") -> bool:
+        pass
+
+
+class CursorField:
+    def __init__(self, path: List[str]) -> None:
+        self._path = path
+
+    def extract_value(self, record: Record) -> Comparable:
+        return _extract_value(record.data, self._path)  # type: ignore  # we assume that the value the path points at is a comparable
+
+
+class Cursor(ABC):
+    @abstractmethod
+    def observe(self, record: Record) -> None:
+        """
+        Indicate to the cursor that the record has been emitted
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def close_partition(self, partition: Partition) -> None:
+        """
+        Indicate to the cursor that the partition has been successfully processed
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def end_sync(self) -> None:
+        """
+        Indicate to the cursor that the sync has been successfully processed
+        """
+        raise NotImplementedError()
+
+
+class NoopCursor(Cursor):
+    def observe(self, record: Record) -> None:
+        pass
+
+    def close_partition(self, partition: Partition) -> None:
+        pass
+
+    def end_sync(self) -> None:
+        pass
+
+
+class ConcurrentCursor(Cursor):
+    _START_BOUNDARY = 0
+    _END_BOUNDARY = 1
+
+    def __init__(
+        self,
+        stream_name: str,
+        stream_namespace: Optional[str],
+        stream_state: Any,
+        message_repository: MessageRepository,
+        connector_state_manager: ConnectorStateManager,
+        cursor_field: CursorField,
+        slice_boundary_fields: Optional[Tuple[str, str]],
```

Comment:
does this need to be comparable or could we enforce it to always be something like `start_datetime` and `end_datetime`?

Reply:
Is this comment referring to `slice_boundary_fields`? Maybe the naming could be improved but those are simple strings that describe where to fetch the cursor values in the slices. 

The part that is extracting the `Comparable` is [here](https://github.com/airbytehq/airbyte/pull/31466/files#diff-995635c52aca1c9b7e47ede7b9e5168b0bb9a2997455362a63bb9d8bae218809R151)